{ibus,fcitx5}-rime: support user provided RIME data packages

linyinfeng opened the pull request:
###### Description of changes

Provide a generic option `i18n.inputMethod.rime.packages` to specify rime data packages used by fcitx5-rime and ibus-rime.

With this PR, third-party rime data packages like [iDvel/rime-ice](https://github.com/iDvel/rime-ice) can be easily packaged and used with ibus-rime and fcitx5-rime.
Also, the `rime-data` package can be refactored to be fine-grained. Currently, nixpkgs only provide a single rime data package called `rime-data`, which contains all official rime data (maintained by the [RIME GitHub organization](https://github.com/rime)). While the Arch Linux community repository provides fine-grained packages like `rime-luna-pinyin`, `rime-double-pinyin`, and `rime-cangjie` etc.

Changes:

1. Add a new option `i18n.inputMethod.rime.packages`. The option itself does not do anything. But It will be used by the fctix5 module and the ibus module to install rime data.
3. Add a new option `i18n.inputMethod.rime.enableDefaultRimeData` for compatibility and out-of-box experience. This option replaces `i18n.inputMethod.fcitx5.enableRimeData`.
4. Update the fcitx5 module to use the new option. Provide `NIX_RIME_DATA_DIR` environment variable in the wrapper instead of adding it to `environment.variables`. The new method should be more robust.
5. Patch ibus-rime to support the `NIX_RIME_DATA_DIR` environment variable (like #167032). And update the ibus module as the fcitx5 module.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Vonfry commented on nixos/modules/i18n/input-method/fcitx5.nix:
```diff
@@ -5,7 +5,7 @@ with lib;
-  addons = cfg.addons ++ optional cfg.enableRimeData pkgs.rime-data;
+  addons = cfg.addons;
   fcitx5Package = pkgs.fcitx5-with-addons.override { inherit addons; };
   whetherRimeDataDir = any (p: p.pname == "fcitx5-rime") cfg.addons;
```
Is it better to move this condition into `config.i18n.inputMethod.rime.enable` and let fcitx5 and ibus add rime addons automatically?

Vonfry replied:
I create [a patch](https://gist.github.com/Vonfry/5cfe2343d72ca02a84a7a91749556ad3) for this suggestion **without test**.

linyinfeng replied:
There are two changes in the patch.

1. This patch moves the `environment.variables, systemPackages, pathsToLink` part from the ibus/fcitx5 modules to the rime module, I think these configs should be defined in the ibus/fcitx5 modules because they are related to the details about how ibus/fcitx5-rime find rime data.
2. I think `config.i18n.inputMethod.rime.enable` is an acceptable change, but it has two disadvantages:
   * Compatibility issue: it's a breaking change to the user's configuration
   * The option will become misleading: users cannot enable rime alone, the option will only work when the proper input method supporting rime is enabled.

   Proper assertions can eliminate the second issue, the main concern is the compatibility issue.

linyinfeng replied:
@Vonfry After reading how the `addons` option of fcitx5 and the `plugins` option of ibus works, how about moving `NIX_RIME_DATA_DIR` to the fcitx5 and ibus wrapper, use the path $out/share/rime-data ($out is the output path of the wrapper).

Vonfry replied:
> moving NIX_RIME_DATA_DIR to the fcitx5 and ibus wrapper

It is better than current method. Thanks.

> The option will become misleading: users cannot enable rime alone.

I never think this problem. You are right. Please forget my suggestion.

linyinfeng committed:
nixos/input-method: support user provided RIME data packages

Provide a generic option i18n.inputMethod.rime.packages to specify
rime data packages used by fcitx5-rime and ibus-rime.

1. Add a new option i18n.inputMethod.rime.packages. It will be used by
the fctix5 module and the ibus module to install rime data.

2. Add a new option i18n.inputMethod.rime.enableDefaultRimeData for
compatibility and out-of-box experience. This option replaces
i18n.inputMethod.fcitx5.enableRimeData.

3. Update the fcitx5 module to use the new option. Provide environment
variable in the wrapper instead of adding it to environment.variables.

4. Patch ibus-rime to support the NIX_RIME_DATA_DIR environment
variable. Update the ibus module and wrapper as fctix5.

linyinfeng commented:
@Vonfry 

> moving NIX_RIME_DATA_DIR to the fcitx5 and ibus wrapper

I have done this. Could you please review it again?

linyinfeng commented:
Tested with ibus on GNOME and fcitx5 on KDE using this flake.
<https://gist.github.com/linyinfeng/fad9e88e02c36197ba82a373758ff258>

Currently, `rime-ice` and `rime-data` have overlap files and cannot be used together, so I tried to package some fine-grained RIME packages in [my nur-packages repository](https://github.com/linyinfeng/nur-packages) for testing.

Vonfry commented:
I think you must write the changelog into [release-note](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md)

linyinfeng commented:
> I think you must write the changelog into [release-note](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md)

Now I have added release notes for this change to this file.

linyinfeng commented:
@ofborg eval

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1941

zendo commented:
Tested all works as expected, lgtm.

oxalica commented:
I strongly prefer #224042 over this. `NIX_RIME_DATA_DIR` is already causing too much trouble for me since #167032. It's hard to update unless reboot (relogin not always work for NixOS). It's impure. It's hard to debug when something went wrong (eg. systemd user services isolates envvars).

Environment variables should be considered harmful.

linyinfeng commented:
https://github.com/NixOS/nixpkgs/pull/224042 has been merged. So I close this one in favor of https://github.com/NixOS/nixpkgs/pull/224042 .
